[Kobo] Enable wake on page turn buttons

kodermike opened the pull request:
Adds menu items (toggles) and functionality to turn a kobo back on with the page turn buttons.

Fix #13668

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/13669)
<!-- Reviewable:end -->

kodermike commented:
fixes #13668

NiLuJe commented on frontend/ui/elements/common_settings_menu_table.lua:
```diff
@@ -237,6 +237,19 @@ if Device:isKobo() then
+
+    if Device:hasKeys() then
+        common_settings.pageturn_power = {
+            text = _("Wakeup if page button pressed"),
```
```suggestion
            text = _("Wake up one page turn key presses"),
```

Commodore64user replied:
We use `page-turn button` elsewhere.

NiLuJe commented on frontend/ui/elements/common_settings_menu_table.lua:
```diff
@@ -237,6 +237,19 @@ if Device:isKobo() then
+            end,
+            callback = function()
+                G_reader_settings:flipNilOrFalse("pageturn_power")
+                UIManager:askForRestart()
```
Unnecessary here

kodermike replied:
Will fix, thanks!

NiLuJe commented:
Incidentally just caught that in a Nickel changelog when updating a device today.

I'm not entirely sure if the page turn keys are actual wakeup sources on older devices with keys, but I would assume so, so, eeeh, why not.

kodermike commented:
> Incidentally just caught that in a Nickel changelog when updating a device today.
> 
> I'm not entirely sure if the page turn keys are actual wakeup sources on older devices with keys, but I would assume so, so, eeeh, why not.

I can verify tonight on a forma (tested this initially on a KLC to see it work, and a Clara2E to see it not break :) )

kodermike commented:
Verified works on the forma, fwiw.

kodermike commented:
> Verified works on the forma, fwiw.

I'm mistaken. Looking into why t doesn't work.

Commodore64user commented on frontend/device/input.lua:
```diff
@@ -822,6 +822,15 @@ function Input:handlePowerManagementOnlyEv(ev)
 
+    -- Treat page turn button like the latest kobo firmware when suspended
+    if G_reader_settings:isTrue("pageturn_power") then
+        if keycode == "RPgBack" or keycode == "LPgBack"
+        or keycode == "RPgFwd" or keycode == "LPgFwd" then
```
It's wild that the `RPg` are on the left and the `LPg` are on the right ;)

kodermike replied:
Have to keep you guessing!

Commodore64user replied:
yeah, don't.

NiLuJe replied:
There's historical and logical reason: page turn buttons are more common on the right side of the device.

Commodore64user commented:
> fixes #13668

This needs to go in the original post https://github.com/koreader/koreader/pull/13669#issue-3016028134

NiLuJe commented:
> > Verified works on the forma, fwiw.
> 
> I'm mistaken. Looking into why t doesn't work.
> 

I assume simply because the page turn buttons are not a wakeup source, so the kernel doesn't even wake the device up.

kodermike commented:
> > I'm mistaken. Looking into why t doesn't work.
> 
> I assume simply because the page turn buttons are not a wakeup source, so the kernel doesn't even wake the device up.

If I had only seen this a few hours ago...but yes, the forma is missing the wakeup-source for the buttons (but has wakeup), which I _think_ explains why the buttons work to wake up in the first minute or so, but not once hibernation has started (anecdotal, but I've noticed the forma isn't completely hibernating after a power button press, probably as it runs through suspension).

Looking now to see if I can figure out which devices support wakeup-source so I can at least tune this proposal a little better. Also to apply it correctly as was pointed out :)

jonnyl2 commented:
I think it's only the KLC that supports it. None of the other devices have any mention of it in the latest FW release notes:

https://pgaskin.net/KoboStuff/kobofirmware.html

NiLuJe commented:
> why the buttons work to wake up in the first minute or so

Yes, suspend entry is delayed by something like 45s (for... ntx reasons).

NiLuJe commented:
> Looking now to see if I can figure out which devices support wakeup-source

Checking the dtb might be a littler overkill (and potentially brittle ;p).

I wouldn't expect the Libra 1/2 to be any different than the Forma, given the board similarities, soo, yeah, I would assume it's MTK-only (for... now?).

NiLuJe commented:
> I wouldn't expect the Libra 1/2 to be any different than the Forma, given the board similarities, soo, yeah, I would assume it's MTK-only (for... now?).

I'll check on the Sage for good measure.

NiLuJe commented:
> I'll check on the Sage for good measure

That would be a no on sunxi:

```
/sys/firmware/devicetree/base/soc@03000000/gpio-keys/
|-- autorepeat
|-- combo
|   |-- gpios
|   |-- label
|   |-- linux,code
|   |-- name
|   `-- wakeup-source
|-- compatible
|-- hall
|   |-- gpios
|   |-- label
|   |-- linux,code
|   |-- name
|   `-- wakeup-source
|-- name
|-- pagedown
|   |-- gpios
|   |-- label
|   |-- linux,code
|   `-- name
|-- pageup
|   |-- gpios
|   |-- label
|   |-- linux,code
|   `-- name
`-- power
    |-- gpios
    |-- label
    |-- linux,code
    |-- name
    `-- wakeup-source
```

So, yeah, gating is behind `isKobo` + `isMTK` ought to do it.

kodermike commented:
Updated label and added condition for `isMTK`. I suspect it's just that the MTK devices have newer kernels and therefore have the ability to set the gpio flags, but at this time new kernel is the same as MTK. In theory someone could compile a newer kernel for their device, or non-kobo's with newer kernel's with gpio compiled in, could also activate this feature. Given that it's currently tied only to one release of kobo, I'd understand if this was rejected. Thanks!

Commodore64user commented on frontend/ui/elements/common_settings_menu_table.lua:
```diff
@@ -237,6 +237,18 @@ if Device:isKobo() then
+
+    if Device:hasKeys() and Device:isMTK() then
+        common_settings.pageturn_power = {
+            text = _("Wake up on Page-turn button presses"),
```
Apologies, lower case "p"

kodermike replied:
> Apologies, lower case "p"

yeah, the one example I looked at had it at the beginning of the sentence. :tableflip:

Commodore64user commented on frontend/ui/elements/common_settings_menu_table.lua:
```diff
@@ -237,6 +237,18 @@ if Device:isKobo() then
+
+    if Device:hasKeys() and Device:isMTK() then
+        common_settings.pageturn_power = {
+            text = _("Wake up on page-turn button presses"),
```
singular, press.

Commodore64user reviewed (approved):
this is my small symbolic attempt to remind @NiLuJe of this PR, has he finally moved out? ;)

NiLuJe commented:
Yeah, I'm slooooowwwly catching up, this is next on my list, ideally this week-end ;).
